src/assets/scripts/client/trafficGenerator/buildPreSpawnAircraft.js

```diff
@@ -0,0 +1,163 @@
+
+    // for each new aircraft
+    for (let i = 0; i < spawnOffsets.length; i++) {
+        let spawnOffset = spawnOffsets[i];
+
+        // for each fix ahead
+        for (let j = 1; j < waypointModelList.length; j++) {
+            const waypoint = waypointModelList[j];
+
+            if (waypoint.distanceFromPreviousWaypoint > spawnOffset) {
+                // if point before next fix
+                const nextFix = waypoint;
+                const previousFix = waypointModelList[j - 1];
+                const heading = bearingToPoint(previousFix.gpsXY, nextFix.gpsXY);
+                const spawnPoint = fixRadialDist(previousFix.gps, heading, spawnOffset);
+                const position = PositionModel.calculatePosition(spawnPoint, airport.position, airport.magnetic_north);
+                // TODO: this looks like it should be a model object
+                const preSpawnHeadingAndPosition = {
+                    heading,
+                    position,
+                    nextFix: nextFix.name
+                };
+
+                spawnPositions.push(preSpawnHeadingAndPosition);
+
+                break;
+            }
+
+            // if point beyond next fix subtract distance from spawnOffset and continue
+            spawnOffset -= waypoint.distanceFromPreviousWaypoint;
+        }
+    }
+
+    return spawnPositions;
+};
+
+/**
+ * Calculate distrance(s) from center where an aircraft should exist onload or airport change
+ *
+ * @function _assembleSpawnOffsets
+ * @param entrailDistance {number}
+ * @param totalDistance {number}
+ * @return spawnOffsets {array<number>}
+ */
+export const _assembleSpawnOffsets = (entrailDistance, totalDistance) => {
+    const spawnOffsets = [];
+
+    // distance between successive arrivals in nm
+    for (let i = entrailDistance; i < totalDistance; i += entrailDistance) {
+        spawnOffsets.push(i);
+    }
+
+    return spawnOffsets;
+};
+
+/**
+ * Calculate heading, nextFix and position data to be used when creating an
+ * `AircraftInstanceModel` along a route.
+ *
+ * @function _preSpawn
+ * @param spawnPatternJson
+ * @param navigationLibrary
+ * @param airport
+ * @return {array<object>}
+ */
+export const _preSpawn = (spawnPatternJson, navigationLibrary, airport) => {
+    // find last fix along STAR that is outside of airspace, ie: next fix is within airspace
+    // distance between closest fix outside airspace and airspace border in nm
+    let extra = 0;
+    let totalDistance = 0;
+    // distance between each arriving aircraft, in nm
+    const entrailDistance = spawnPatternJson.speed / spawnPatternJson.rate;
+    const activeRouteModel = new RouteModel(spawnPatternJson.route);
+    const isPreSpawn = true;
+    const waypointModelList = airport.findWaypointModelsForStar(
+        activeRouteModel.procedure,
+        activeRouteModel.entry,
+        airport.runway,
+        isPreSpawn
+    );
+
+    for (let i = 0; i < waypointModelList.length; i++) {
```

Comment:
easy abstraction here on this for loop (maybe to `_calculateDistancesAlongRoute`?). And if we return an object, it gives a wonderful excuse to destructure:

`const { totalDistance, extra } = _calculateDistancesAlongRoute(waypointModelList, airport);`

\* ooooh aaaah \*